src/foam/core/FObject.java

```diff
@@ -302,7 +302,62 @@ default FObject copyFrom(FObject obj) {
           if ( p2 != null ) {
             if ( p2.isSet(obj) ) p.set(this, p2.get(obj));
           }
-        } catch (ClassCastException ignore) {}
+        } catch (ClassCastException ignore) {
+          System.err.println("FObject.copyFrom "+p.getName()+" "+ignore.getMessage());
+        }
+      }
+    }
+    return this;
+  }
+
+  /**
+   * Similar to copyFrom, with recursion on nested FObjects, only setting
+   * isSet=true properties.
+   */
+  default FObject overlay(FObject obj) {
+    return overlay_(obj, new java.util.HashSet());
+  }
+
+  default FObject overlay_(FObject obj, java.util.Set visited) {
+    int code = obj.hashCode();
+    if ( visited.contains(code) ) {
+      return this;
+    } else {
+      visited.add(code);
+      if ( this.hashCode() == obj.hashCode() ) {
+        return this;
+      }
+    }
+
+    List<PropertyInfo> props = obj.getClassInfo().getAxiomsByClass(PropertyInfo.class);
+    for ( PropertyInfo p : props ) {
+      Object remote = null;
+      try {
+        remote = p.get(obj);
+      } catch ( ClassCastException e ) {
+        PropertyInfo p2 = (PropertyInfo) getClassInfo().getAxiomByName(p.getName());
+        if ( p2 != null ) {
+          p = p2;
+          try {
+            remote = p.get(obj);
+          } catch ( ClassCastException ee ) {
+            System.err.println("FObject.overlay "+p.getName()+" get "+ee);
+          }
+        }
+      }
+      try {
+        if ( p.isSet(obj) ) {
+          Object local = p.get(this);
+          if ( remote instanceof FObject &&
+               local != null &&
+               ! local.equals(remote) ) {
```

Comment:
if remote is null then the instanceof check would return false otherwise we could safely check if it equals to local.
```suggestion
          if ( remote instanceof FObject &&
               ! remote.equals(local) ) {
```

Reply:
Yes, I understand, but we either have to check remote for null or local for null, so pick one.

Reply:
ah I see, sorry for confusion.